### PR TITLE
ARROW-10742: [Python] Check mask when creating array from numpy

### PR DESCRIPTION
**Pull request by chrisavl**
This change adds checks so that the same exceptions are raised when creating an array with a mask using a python sequence or a numpy array.

### CODE REVIEW

**Review by jorisvandenbossche — COMMENTED** [review]
@chrisavl thanks for the issue report and PR!

Looks good to me.

The issue is that the C++ `NdarrayToArrow` doesn't do any validation of the mask (it only checks that the values array is a 1D numpy array, but doesn't do the same for the mask array), while `ConvertPySequence` actually checks the mask array (in `VisitSequenceMasked`). 
@pitrou is there a preference to also add this check to the C++ function, or is the check on the Python side as this PR fine?

File: python/pyarrow/array.pxi [hunk 1]
```diff
@@ -238,6 +238,15 @@ def array(object obj, type=None, mask=None, size=None, from_pandas=None,
                 mask = None if values.mask is np.ma.nomask else values.mask
                 values = values.data
 
+        if mask is not None:
+            if mask.dtype != np.bool_:
+                raise ValueError("Mask must be boolean dtype")
```
Comment (pitrou) [inline python/pyarrow/array.pxi:243]: Make this one a TypeError?

**Review by pitrou — COMMENTED** [review]
Thanks for doing this. Just a nit.

File: python/pyarrow/array.pxi (hunk 1 above)
Comment (chrisavl) [reply python/pyarrow/array.pxi]: Done!

**Comment by jorisvandenbossche** [comment]
Thanks @chrisavl !

**Pull request closed without merging**